pkg/storegateway/bucket_stores.go

```diff
@@ -151,11 +152,31 @@ func (u *BucketStores) InitialSync(ctx context.Context) error {
 
 // SyncBlocks synchronizes the stores state with the Bucket store for every user.
 func (u *BucketStores) SyncBlocks(ctx context.Context) error {
-	return u.syncUsersBlocks(ctx, func(ctx context.Context, s *store.BucketStore) error {
+	return u.syncUsersBlocksWithRetries(ctx, func(ctx context.Context, s *store.BucketStore) error {
 		return s.SyncBlocks(ctx)
 	})
 }
 
+func (u *BucketStores) syncUsersBlocksWithRetries(ctx context.Context, f func(context.Context, *store.BucketStore) error) error {
+	retries := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 10 * time.Second,
+		MaxRetries: 3,
+	})
+
+	var lastErr error
+	for retries.Ongoing() {
+		lastErr = u.syncUsersBlocks(ctx, f)
+		if lastErr == nil {
+			return nil
+		}
+
+		retries.Wait()
+	}
+
+	return lastErr
```

Comment:
This can return `nil` (which means success!) if context was cancelled before first call to `retries.Ongoing()` and thus `u.syncUsersBlocks(ctx, f)`. We should add a check for that -- if `lastErr` is `nil`, we need to return `retries.Err()` instead.

(I think we should extend `Backoff` to remember failures, and then it's Err() method could return last failure, and simplify client code. But we can do that in separate PR)

Reply:
Just to clarify, you're proposing something like this:

```golang
if lastErr == nil {
   return retries.Err()
}

return lastErr
```
Right?

Reply:
Yes, that is what I mean.